### PR TITLE
Mesh Generator infrastructure changes to support --csg-only command line option

### PR DESCRIPTION
**Pull request by shikhar413**
refs #29095

### CODE REVIEW

**Comment by shikhar413** [comment]
@kkiesling @eshemon FYI

**Comment by shikhar413** [comment]
@GiudGiud @loganharbour this is a preliminary PR that defines the `--csg-only` command line option for triggering CSG-based mesh generation. We (Kalin and I) figured this would be the best way to submit an intermediate PR. Trying to add the functionality for defining a CSGBase class object is not in this PR because it would become too large to review, so I've left some TODO's of areas that need to be updated. We've tried to leverage the work from last year's FY that invokes data driven generation for this work as well, but we can schedule a separate meeting to discuss this code design

**Comment by loganharbour** [comment]
I'm not sure that it makes sense to add this yet without the actual storage of the CSG representation? Has that component been discussed?

**Comment by shikhar413** [comment]
Thought I’d push this up first to get your thoughts on whether this could be merged first before creating the data type for storage of CSG representation, since adding that data type will be quite a large number of changes as well. That portion has not been completed yet but if you’d like that part to be included in this PR as well we can treat this PR as a draft and you guys can review the code design so far. Let us know what makes most sense in terms of code review

**Comment by shikhar413** [comment]
Just an update on this. I think I've figured out a way to generate a CSG mesh from any arbitrary Exodus file that is imported through FileMeshGenerator, the intermediate goal for this PR could then be to demonstrate `--csg-only` option + `generateCSG` works for `FileMeshGenerator`. Would that be sufficient for now? This way we can define the CSGBase object that is needed to get this running and go from there. Just as an example of what I have so far:

image

**Comment by GiudGiud** [comment]
Adding CSG output for a mesh generator would be great to demonstrate this CSG-only mode.
FMG is the more complex one but also one of the more useful so feel free to do that one if you have a good lead!

what is your target output format? XML?

**Comment by shikhar413** [comment]
> what is your target output format? XML?
how are you conserving subdomains of elements in the mesh? Using universes / cells containing cells?

right now targeting json output format so as not to favor one MC code over another. Then the downstream code (Cardinal / Shift etc) can have an in memory hook to convert this json data to whatever format they need

**Comment by GiudGiud** [comment]
Do you expect you'll output some explanation of the CSG through comments? This would make json a problem.

**Comment by GiudGiud** [comment]
I dont know enough. I just asked Paul on the crpg slack let's see what he thinks on formats

**Comment by shikhar413** [comment]
I’m honestly not too concerned with what the output data type is going to be here, as long as it contains the data that is needed to generate the CSG geometry. Ultimately I see this more of a debugging tool to make sure CSG-only is working as expected, and whatever format we choose we’re going to have to modify it to fit whatever the downstream MC code can read. We also might want to discuss whether we are creating a geometry specification file or a geometry input file (e.g. for OpenMC an XML file or a Python file?)

**Comment by GiudGiud** [comment]
> Ultimately I see this more of a debugging tool to make sure CSG-only is working as expected

Yes debugging and testing as well. I think that will be the main way of testing the CSG generation in MOOSE, which is why I'm happy to see this capability come in quite early.

> We also might want to discuss whether we are creating a geometry specification file or a geometry input file (e.g. for OpenMC an XML file or a Python file?

This would be done in the OpenMC repo right? With a MOOSE-JSON-CSV reader/converter python script?

**Comment by shikhar413** [comment]
> This would be done in the OpenMC repo right? With a MOOSE-JSON-CSV reader/converter python script?

Yup, or probably Cardinal since I don't think OpenMC directly interfaces with finite element meshes. Will have to coordinate with the external stakeholders for MC codes to figure this step out, but for now we can focus on the FEM-->CSG JSON step (or another output format if there is a strong reason to do so)

**Comment by shikhar413** [comment]
The other thing to also note is that if we are coupling with a C++ code we can ignore the output data type altogether and generate the final CSG mesh from the internal representation we create in MOOSE

File: framework/include/actions/CSGOnlyAction.h [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "Action.h"
+
+class CSGOnlyAction : public Action
```
Comment (GiudGiud) [inline framework/include/actions/CSGOnlyAction.h:14]:
```suggestion
/**
 * Outputs the Constructive Solid Geometry to file then exits
 */
class CSGOnlyAction : public Action
```

File: framework/include/meshgenerators/MeshGenerator.h [hunk 2]
```diff
@@ -195,6 +220,11 @@ class MeshGenerator : public MooseObject, public MeshMetaDataInterface
    */
   virtual void generateData();
 
+  /**
+   * Generate the CSG mesh
```
Comment (GiudGiud) [inline framework/include/meshgenerators/MeshGenerator.h:224]:
```suggestion
   * Generate the CSG representation of the mesh (or meshing operation) created by this mesh generator
```

File: framework/src/base/MooseApp.C [hunk 3]
```diff
@@ -108,6 +108,11 @@ MooseApp::validParams()
       "mesh_only",
       "--mesh-only [mesh_file_name]",
       "Setup and Output the input mesh only (Default: \"<input_file_name>_in.e\")");
+  // TODO: csg_only option will need to be updated to reflect output file format for CSG base file
+  params.addCommandLineParam<std::string>(
+      "csg_only",
+      "--csg-only",
+      "Setup and Output the input mesh in CSG format only (Default: \"<input_file_name>_in.e\")");
```
Comment (GiudGiud) [inline framework/src/base/MooseApp.C:115]:
```suggestion
      "Setup and Output a constructive solid geometry (CSG) representation of the input mesh (Default: \"<input_file_name>_in.json\")");
```

File: framework/src/meshgenerators/MeshGenerator.C [hunk 4]
```diff
@@ -326,6 +339,13 @@ MeshGenerator::generateInternal()
   return mesh;
 }
 
+void
+MeshGenerator::generateInternalCSG()
+{
+  mooseAssert(isDataOnly(), "Trying to use csg-only mode while not in data-driven mode");
```
Comment (GiudGiud) [inline framework/src/meshgenerators/MeshGenerator.C:345]: ?

**Review by GiudGiud — COMMENTED** [review]
quick comments because I was curious. Looking forward to seeing how you make the CSGBase work

**Comment by shikhar413** [comment]
@GiudGiud @loganharbour just pushed where I'm at with integrating --csg-only with `FileMeshGenerator`. Everything looks like it's working as expected, just a few minor things left such as adding docstrings, documentation, and some more unit tests, as well as updating the logic for handling `CSGBase` objects in the `CSGMeshOnlyAction`. The `CSGBase` class is defined now, and I have a function CSGBase::generateOutput() which defines the OpenMC Python API input from Exodus mesh like so:

```
surf0 = openmc.Plane(name='surf0', a=-0.57735, b=1, c=0, d=0)
surf1 = openmc.Plane(name='surf1', a=-0.57735, b=-1, c=0, d=-1.1547, boundary_type='vacuum')
surf2 = openmc.Plane(name='surf2', a=1.1547, b=0, c=0, d=0)
surf3 = openmc.Plane(name='surf3', a=0.57735, b=1, c=-8.5876e-18, d=0)
surf4 = openmc.Plane(name='surf4', a=-1.1547, b=-2.22045e-16, c=-1.28198e-16, d=-1.1547, boundary_type='vacuum')
surf5 = openmc.Plane(name='surf5', a=1.1547, b=1.4141e-16, c=-1.40149e-33, d=0)
surf6 = openmc.Plane(name='surf6', a=-0.57735, b=1, c=8.5876e-18, d=-1.1547, boundary_type='vacuum')
surf7 = openmc.Plane(name='surf7', a=0.57735, b=-1, c=-8.5876e-18, d=0)
surf8 = openmc.Plane(name='surf8', a=0.57735, b=1, c=-2.34618e-17, d=-1.1547, boundary_type='vacuum')
surf9 = openmc.Plane(name='surf9', a=-0.57735, b=-1, c=-2.57628e-17, d=-0)
surf10 = openmc.Plane(name='surf10', a=1.1547, b=8.88178e-16, c=0, d=-1.1547, boundary_type='vacuum')
surf11 = openmc.Plane(name='surf11', a=0.57735, b=-1, c=2.57628e-17, d=-1.1547, boundary_type='vacuum')
void_cell = openmc.Cell(name='void_cell', fill=None, region=-surf1 & -surf4 & -surf6 & -surf8 & -surf10 & -surf11)
cell0 = openmc.Cell(name='cell0', fill=hexagon, region=+surf0 & +surf1 & +surf2)
cell1 = openmc.Cell(name='cell1', fill=hexagon, region=-surf0 & +surf3 & +surf4)
cell5 = openmc.Cell(name='cell5', fill=hexagon, region=-surf2 & -surf9 & +surf11)
cell2 = openmc.Cell(name='cell2', fill=hexagon, region=-surf3 & +surf5 & +surf6)
cell3 = openmc.Cell(name='cell3', fill=hexagon, region=-surf5 & +surf7 & +surf8)
cell4 = openmc.Cell(name='cell4', fill=hexagon, region=-surf7 & +surf9 & +surf10)
geometry = openmc.Geometry([void_cell, cell0, cell1, cell5, cell2, cell3, cell4])
```

will need to update this to generate a JSON file instead. I'll let you know when the PR is ready to remove from being in draft stage

**Comment by shikhar413** [comment]
@loganharbour @loganharbour after discussing with @kkiesling, in order to keep the code changes more focused we will close this PR for now, and instead submit a PR that only has the `--csg-only` functionality as well as a more fleshed out version of the `CSGBase` class. The specific implementation of FileMeshGenerator that calls on the CSGBase class will come in a follow-on PR.

**Pull request closed without merging**